src/main/java/org/eclipse/yasson/internal/serializer/types/EnumSerializer.java

```diff
@@ -12,26 +12,57 @@
 
 package org.eclipse.yasson.internal.serializer.types;
 
+import java.util.EnumMap;
+
 import jakarta.json.stream.JsonGenerator;
 
 import org.eclipse.yasson.internal.SerializationContextImpl;
+import org.eclipse.yasson.internal.model.ClassModel;
+import org.eclipse.yasson.internal.model.PropertyModel;
+
 
 /**
  * Serializer of the {@link Enum} types.
  */
 class EnumSerializer extends TypeSerializer<Enum<?>> {
 
+    private final EnumMap<? extends Enum<?>, String> constantToNameMap;
+
     EnumSerializer(TypeSerializerBuilder serializerBuilder) {
         super(serializerBuilder);
+
+        constantToNameMap = createConstantToNameMap(serializerBuilder);
+    }
+
+    private static <E extends Enum<E>> EnumMap<E, String> createConstantToNameMap(TypeSerializerBuilder serializerBuilder) {
+        EnumMap<E, String> constantToNameMap = null;
+        Class<?> clazz = serializerBuilder.getClazz();
+
+        if (clazz.isEnum()) {
+            try {
+                @SuppressWarnings("unchecked")
+                Class<E> enumClazz = (Class<E>) clazz;
+                constantToNameMap = new EnumMap<>(enumClazz);
+                ClassModel classModel = serializerBuilder.getJsonbContext().getMappingContext().getOrCreateClassModel(clazz);
+
+                for (E enumConstant : enumClazz.getEnumConstants()) {
+                    PropertyModel model = classModel.getPropertyModel(enumConstant.name());
+                    constantToNameMap.put(enumConstant, model.getWriteName());
+                }
+            } catch (ClassCastException classCastException) {
+                throw new IllegalArgumentException("EnumSerializer can only be used with Enum types");
+            }
+        }
+        return constantToNameMap;
     }
 
     @Override
     void serializeValue(Enum<?> value, JsonGenerator generator, SerializationContextImpl context) {
-        generator.write(value.name());
+        generator.write(constantToNameMap == null ? value.name() : constantToNameMap.get(value));
```

Comment:
the same as wit Deserializer. When this map can be null?

Reply:
I already answered this above.